marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/impl/sm-es-impl.xqy

```diff
@@ -137,3 +137,102 @@ declare function es-impl:get-entity-def-property(
             fn:error($const:ENTITY-PROPERTY-NOT-FOUND-ERROR, ("Specified entity property not found"), ($entity-def, $property-title))
   else ()
 };
+
+declare variable $indexes-by-property-names := map:map();
+
+declare function es-impl:get-entity-property-info($entity-type-iri as xs:string, $property-path as xs:string) as map:map
+{
+  let $property-key := $entity-type-iri || "|" || $property-path
+  let $property-info := es-impl:get-entity-property-info() => map:get($property-key)
+  return (
+    $property-info,
+    if (xdmp:trace-enabled($const:TRACE-MATCH-RESULTS)) then
+      xdmp:trace($const:TRACE-MATCH-RESULTS, "Retrieving property information '"|| $property-key ||"' : " || xdmp:to-json-string($property-info))
+    else ()
+  )
+};
+
+declare function es-impl:get-entity-property-info() as map:map
+{
+  if (map:count($indexes-by-property-names) = 0) then
+    let $element-index-iri := sem:iri('http://marklogic.com/entity-services#ElementRangeIndexedProperty')
+    let $path-index-iri := sem:iri('http://marklogic.com/entity-services#PathRangeIndexedProperty')
+    let $properties := sem:sparql('
```

Comment:
Pinging @fsnow about this - what's the benefit of using sparql to get properties vs just using cts.doc to get the entity document? I feel that any performance benefit would be slight here, as entity docs are likely to be hanging around in the expanded tree cache most of the time. And the downside is that most backend developers can understand code that uses JS to grab property info, whereas very few developers are conversant enough in sparql to understand the query below - in particular, how to enhance it in the future as the need arises.

Reply:
The entity properties data is likely to be in cache for either approach, expanded tree cache vs triple cache, so the performance should be roughly equivalent. I would avoid introducing SPARQL queries in the data hub code for reasons that Rob gives above: developer familiarity.